packages/ra-core/src/dataProvider/useCreate.ts

```diff
@@ -80,12 +92,111 @@ export const useCreate = <
+    // when the calling component is unmounted
+    const callTimeOnError =
+        useRef<
+            UseCreateOptions<
+                RecordType,
+                MutationError,
+                ResultRecordType
+            >['onError']
+        >();
+    const callTimeOnSettled =
+        useRef<
+            UseCreateOptions<
+                RecordType,
+                MutationError,
+                ResultRecordType
+            >['onSettled']
+        >();
+
+    // We don't need to keep a ref on the onSuccess callback as we call it ourselves for optimistic and
+    // undoable mutations. There is a limitation though: if one of the side effects applied by the onSuccess callback
+    // unmounts the component that called the useUpdate hook (redirect for instance), it must be the last one applied,
+    // otherwise the other side effects may not applied.
     const hasCallTimeOnSuccess = useRef(false);
-    const hasCallTimeOnSettled = useRef(false);
-    const { getMutateWithMiddlewares, ...mutationOptions } = options;
+
+    const updateCache = ({ resource, id, data, meta }) => {
+        // hack: only way to tell react-query not to fetch this query for the next 5 seconds
+        // because setQueryData doesn't accept a stale time option
+        const now = Date.now();
+        const updatedAt = mode.current === 'undoable' ? now + 5 * 1000 : now;
+        // Stringify and parse the data to remove undefined values.
+        // If we don't do this, an update with { id: undefined } as payload
+        // would remove the id from the record, which no real data provider does.
+        const clonedData = JSON.parse(JSON.stringify(data));
+
+        const updateColl = (old: RecordType[]) => {
+            if (!old) return old;
+            return [data, ...old];
+        };
+
+        type GetListResult = Omit<OriginalGetListResult, 'data'> & {
+            data?: RecordType[];
+        };
+
+        queryClient.setQueryData(
+            [resource, 'getOne', { id: String(id), meta }],
+            (record: RecordType) => ({ ...record, ...clonedData }),
+            { updatedAt }
+        );
+        queryClient.setQueriesData(
+            { queryKey: [resource, 'getList'] },
+            (res: GetListResult) =>
+                res && res.data ? { ...res, data: updateColl(res.data) } : res,
+            { updatedAt }
+        );
+        queryClient.setQueriesData(
+            { queryKey: [resource, 'getInfiniteList'] },
+            (
+                res: UseInfiniteQueryResult<
+                    InfiniteData<GetInfiniteListResult>
+                >['data']
+            ) =>
+                res && res.pages
+                    ? {
+                          ...res,
+                          pages: res.pages.map(page => ({
+                              ...page,
+                              data: updateColl(page.data),
+                          })),
+                      }
+                    : res,
```

Comment:
Are you adding the new record to all pages of the infinite query? Won't that lead to the record appearing multiple times in the list? :thinking: